internal/engine/upper.go

```diff
@@ -76,27 +76,30 @@ func (u Upper) CreateServices(ctx context.Context, services []model.Service, wat
 			case <-ctx.Done():
 				return ctx.Err()
 			case event := <-sw.events:
+				oldState := buildStates[event.service.Name]
+				buildState := oldState.NewStateWithFilesChanged(event.files)
+				buildStates[event.service.Name] = buildState
+
+				changedFiles := buildState.FilesChanged()
 				var changedPathsToPrint []string
-				if len(event.files) > maxChangedFilesToPrint {
-					changedPathsToPrint = append(changedPathsToPrint, event.files[:maxChangedFilesToPrint]...)
+				if len(changedFiles) > maxChangedFilesToPrint {
+					changedPathsToPrint = append(changedPathsToPrint, changedFiles[:maxChangedFilesToPrint]...)
 					changedPathsToPrint = append(changedPathsToPrint, "...")
 				} else {
-					changedPathsToPrint = event.files
+					changedPathsToPrint = changedFiles
 				}
 
 				logger.Get(ctx).Infof("files changed. rebuilding %v. observed %d changes: %v",
-					event.service.Name, len(event.files), ospath.TryAsCwdChildren(changedPathsToPrint))
+					event.service.Name, len(changedFiles), ospath.TryAsCwdChildren(changedPathsToPrint))
 
-				var err error
-				token, err := u.b.BuildAndDeploy(
+				result, err := u.b.BuildAndDeploy(
 					ctx,
 					event.service,
-					buildTokens[event.service.Name],
-					event.files)
+					buildState)
 				if err != nil {
 					logger.Get(ctx).Infof("build failed: %v", err.Error())
```

Comment:
aaaah i see, because we only set a new build state from the result if err == nil, this way we retain the changed files from the last time through this loop if the build fails and we have to try again, yeah?

Reply:
yup! also added a test for this case, which is the main motivation for this pr